flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamTask.java

```diff
@@ -733,8 +764,7 @@ void restoreInternal() throws Exception {
         // needed
         channelIOExecutor.shutdown();
 
-        isRunning = true;
-        isRestoring = false;
+        taskState.status = TaskState.Status.RUNNING;
```

Comment:
At this point, the task might be already cancelled, right?

Previously, this was captured by `isRunning = true` and `isCancelled = true`.
But now, we overwrite `CANCELLED` by `RUNNING`.

This might break checks like [this one](https://github.com/apache/flink/blob/master/flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/SourceStreamTask.java#L339) in `SourceStreamTask` and cause task and then job failure.

---

I guess this is basically the same concern expressed previously that we might lose some transitions.
I see only two ways to be on the safe side:
1. Have all the same flags, but inside the inner `TaskState` class
1. Model the enum exactly as `2^number_of_flags` AND transition state via CAS (not just assignment)

WDYT?

Reply:
Hey Roman, very good point since TaskCanceler is asynchronous in nature.
However, the only way to invoke [Task cancelation](https://github.com/apache/flink/blob/8be94e6663d8ac6e3d74bf4cd5f540cc96c8289e/flink-runtime/src/main/java/org/apache/flink/runtime/taskmanager/Task.java#L1662) is through the [invokable method](https://github.com/apache/flink/pull/21923/files#diff-0c5fe245445b932fa83fdaf5c4802dbe671b73e8d76f39058c6eaaaffd9639faL982-L983) that simultaneously marks the task as **not** Running and Canceled. 

In that sense, its not the same situation as failing, so I believe we are safe here.

Reply:
Hey Panos,
> the [invokable method](https://github.com/apache/flink/pull/21923/files#diff-0c5fe245445b932fa83fdaf5c4802dbe671b73e8d76f39058c6eaaaffd9639faL982-L983) that simultaneously marks the task as not Running and Canceled

Yes, but after that, the Task Thread might update `isRunning`, but not `cancelled`.

1. Task Thread enters `restoreInternal` and passes `ensureNotCanceled`
2. Canceller Thread calls `invokable.cancel` and sets `isRunning = false`, `canceled = true`
3. Task Thread continues and sets `isRunning = true`, `isRestoring = false`
4. Legacy Source Thread gets interrupted and checks `isCancelled` - it's `true`, so `CancelTaskException` is thrown, which won't fail the job

With this PR, all the flags are combined and updated unconditionally:
1. Task Thread enters `restoreInternal` and passes `ensureNotCanceled`
2. Canceller Thread calls `invokable.cancel` and sets `taskState.status = TaskState.Status.CANCELED`
3. Task Thread continues and sets `taskState.status = TaskState.Status.RUNNING`
4. Legacy Source Thread gets interrupted and checks `isCancelled`, i.e. `taskState.status == TaskState.Status.CANCELED`; which is `false`, so some other `Throwable` is thrown, which **does** fail the job

Reply:
As I understand right now all transitions are correct. The only problem is we can overwrite the final status with another final status which is incorrect behaviour. To fix that, I agree we can use CAS to update TaskStateStatus and allow only transitions described in javadoc.

I also don't think that we should get back to 4 flags since we already spend a lot of time understanding which combinations of them are really possible and it seems nobody knows for sure. So it is better to decrease these combinations and I think the enum with clear transitions is good choice.

I still think that we should even get rid of `failing` flag since I feel we can combine it with enum but I'm ok if we do it later

Reply:
> now all transitions are correct

I'd add CANCELLING state to reflect the above combination of flags.

> I also don't think that we should get back to 4 flags since we already spend a lot of time understanding which combinations of them are really possible

Makes sense. I'm not insisting on any option. As for the knowledge, should we encode it as allowed state transitions? I believe it was there previously in the PR, but then was deleted. 

> I still think that we should even get rid of failing

👍 and it seems just necessary if we proceed with CAS

Reply:
Potentially we can have CANCELLING status but right now I don't see where we can actually use it. For example, if we consider the above-described scenario, we have a situation when canceled=true at the same time as isRestoring=true as well. isRestoring we use only inside `handleAsyncException` and if we want to keep this logic we indeed need some extra status like CANCELLING but I actually think that we don't need to keep that logic since if the job is canceled already we can easily ignore other exceptions which can happen since they can be actually because of cancelation. I have the same thoughts about canceled=true + isRunning=true.(I don't see where we should execute logic isRunning=true if the canceled is true already.)
Long story short, if we indeed understand where we need CANCELLING status let's add it but if we need it only for keeping the old behavior let's discuss this old behavior since I am not really sure that the old behavior is correct.

Reply:
> we don't need to keep that logic since if the job is canceled already we can easily ignore other exceptions which can happen since they can be actually because of cancelation

Please note that Task cancellation is not the same as Job cancellation. For example, in case of some intermittent failure, the Job will be `RESTARTING`, but the tasks will be cancelled and then rescheduled. Ideally, we shouldn't be ignoring **any** exception during restart (just interruptions and cancellations). Otherwise, it might create a resource leak (job correctness won't be affected though).

As for the `running` flag (which will report `true` on `master`; and `false` when CAS to `CANCELLED`);
I briefly skimmed though its usages:
- `Task.restoreAndInvoke` -> `StreamTask.invoke` -> `StreamTask.restoreInternal` - regression: it will allow 2nd `restoreInternal`
- ongoing checkpoints will not be aborted by RPC - that's fine
- local state will not be pruned by CHK abort RPC - regression (but **likely** will be pruned on the next checkpoint)
- ...

The above can probably be fixed (in this PR), but my point is that it's hard to check all the combinations and their usages.
If that's true, I'd stick closer to the original behavior by having states **unless** we are sure that they are impossible.